src/telempostdaemon.c

```diff
@@ -210,11 +212,96 @@ size_t write_callback(char *ptr, size_t size, size_t nmemb, void *userdata)
+        char *key = NULL;
+        char *value = NULL;
+        json_object *value_json[NUM_HEADERS + 1] = {NULL}; // Extra index for payload.
+        json_object *json_record = NULL;
+        char *json_str = NULL;
+
+        json_record = json_object_new_object(); // Create new JSON object.
+
+        for (int i = 0; i < NUM_HEADERS; i++)
+        {
+                size_t header_len = strlen(headers[i]);
+                char *tmp_header = strndup(headers[i], header_len);
+                char *tok = strtok(tmp_header, ":");
+
+                telem_log(LOG_DEBUG, "headers[%d] = \"%s\"\n", i, headers[i]);
+
+                key = strdup(tok);                        // Key string.
+
+                tok = strtok(NULL, "\n");
+                tok++;                                    // Skip the space.
+
+                value = strdup(tok);                      // Value string.
+
+                // Create the appropriate JSON object.
+                if ((strcmp(key, TM_RECORD_FORMAT_VERSION_STR) == 0) ||
+                    (strcmp(key, TM_SEVERITY_STR)       == 0) ||
+                    (strcmp(key, TM_PAYLOAD_FORMAT_VERSION_STR) == 0))
+                {
+                        // Create JSON value of type int32.
+                        value_json[i] = json_object_new_int(atoi(value));
+                }
+                else if (strcmp(key, TM_CREATION_TIMESTAMP_STR) == 0){
+                        // Create JSON value of type int64.
+                        value_json[i] = json_object_new_int64(atoi(value));
+                }
+                else
+                {
+                        // Create JSON value of type string.
+                        value_json[i] = json_object_new_string(value);
+                }
+
+                // Add key-string/value-string pair to JSON object.
+                result = json_object_object_add(json_record, key, value_json[i]);
+
+                // Free current header, key and value strings to process the next
+                // record header.
+                free(tmp_header);
+                free(key);
+                free(value);
+
+                if (result != 0)
+                {
+                        // Could not add key/value pair to JSON object.
+                        // Fail. json_str is NULL at this point.
+                        goto cleanup;
+                }
+        }
+
+        telem_log(LOG_DEBUG, "payload = \"%s\"\n", payload);
+
+        // Add payload key/value pair to JSON object.
+        value_json[NUM_HEADERS] = json_object_new_string(payload);
+        result = json_object_object_add(json_record, "payload",
+                                        value_json[NUM_HEADERS]);
+
+        if (result != 0)
+        {
+                // Could not add key/value pair to JSON object.
+                // Fail. json_str is NULL at this point.
+                goto cleanup;
+        }
+
+        json_str = strdup(json_object_to_json_string(json_record));
```

Comment:
Although I agree with you in principle, the JSON library is limited. JSON objects are "special" and can only be freed with a call to json_object_put(). So wrote the code to keep JSON API specific allocation-freeing in the same function scope. Because of this I had to duplicate the string with strdup(). I don't like duplicating memory contents and passing them around. As you allude to, passing references to string is more memory-efficient, but we are limited by JSON's special way of creating string objects.